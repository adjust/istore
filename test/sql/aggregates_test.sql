BEGIN;
-- isagg int should skip null keys;
-- ./spec/istore/aggregates_spec.rb:9;
CREATE EXTENSION istore;
SELECT id, isagg(NULLIF(i%10,3), NULLIF(i::int, 50) ) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
ROLLBACK;
BEGIN;
-- isagg int should skip null values;
-- ./spec/istore/aggregates_spec.rb:16;
CREATE EXTENSION istore;
SELECT id, isagg((i%10), NULL::int) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
ROLLBACK;
BEGIN;
-- isagg bigint should skip null keys;
-- ./spec/istore/aggregates_spec.rb:9;
CREATE EXTENSION istore;
SELECT id, isagg(NULLIF(i%10,3), NULLIF(i::bigint, 50) ) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
ROLLBACK;
BEGIN;
-- isagg bigint should skip null values;
-- ./spec/istore/aggregates_spec.rb:16;
CREATE EXTENSION istore;
SELECT id, isagg((i%10), NULL::bigint) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
ROLLBACK;
