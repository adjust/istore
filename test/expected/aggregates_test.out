BEGIN;
-- isagg int should skip null keys;
-- ./spec/istore/aggregates_spec.rb:9;
CREATE EXTENSION istore;
SELECT id, isagg(NULLIF(i%10,3), NULLIF(i::int, 50) ) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
 id |                                                   isagg                                                    
----+------------------------------------------------------------------------------------------------------------
  1 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
  2 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
  3 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
(3 rows)

ROLLBACK;
BEGIN;
-- isagg int should skip null values;
-- ./spec/istore/aggregates_spec.rb:16;
CREATE EXTENSION istore;
SELECT id, isagg((i%10), NULL::int) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
 id | isagg 
----+-------
  1 | 
  2 | 
  3 | 
(3 rows)

ROLLBACK;
BEGIN;
-- isagg bigint should skip null keys;
-- ./spec/istore/aggregates_spec.rb:9;
CREATE EXTENSION istore;
SELECT id, isagg(NULLIF(i%10,3), NULLIF(i::bigint, 50) ) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
 id |                                                   isagg                                                    
----+------------------------------------------------------------------------------------------------------------
  1 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
  2 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
  3 | "0"=>"500", "1"=>"460", "2"=>"470", "4"=>"490", "5"=>"500", "6"=>"510", "7"=>"520", "8"=>"530", "9"=>"540"
(3 rows)

ROLLBACK;
BEGIN;
-- isagg bigint should skip null values;
-- ./spec/istore/aggregates_spec.rb:16;
CREATE EXTENSION istore;
SELECT id, isagg((i%10), NULL::bigint) FROM generate_series(1,100) i, generate_series(1,3) id GROUP BY id ORDER BY id;;
 id | isagg 
----+-------
  1 | 
  2 | 
  3 | 
(3 rows)

ROLLBACK;
BEGIN;
-- extend istore should correctly extend internal aggregation state while summing istores which total length is larger than default capacity;
-- ./spec/istore/aggregates_spec.rb:32;
CREATE EXTENSION istore;
CREATE TABLE test (v istore);
INSERT INTO test SELECT istore(array_agg(x), array_agg(x)) FROM generate_series(1, 31, 2) AS a(x);
INSERT INTO test SELECT istore(array_agg(x), array_agg(x)) FROM generate_series(2, 32, 2) AS a(x);
SELECT SUM(v) FROM test;
                                                                                                                                                                                     sum                                                                                                                                                                                      
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 "1"=>"1", "2"=>"2", "3"=>"3", "4"=>"4", "5"=>"5", "6"=>"6", "7"=>"7", "8"=>"8", "9"=>"9", "10"=>"10", "11"=>"11", "12"=>"12", "13"=>"13", "14"=>"14", "15"=>"15", "16"=>"16", "17"=>"17", "18"=>"18", "19"=>"19", "20"=>"20", "21"=>"21", "22"=>"22", "23"=>"23", "24"=>"24", "25"=>"25", "26"=>"26", "27"=>"27", "28"=>"28", "29"=>"29", "30"=>"30", "31"=>"31", "32"=>"32"
(1 row)

ROLLBACK;
BEGIN;
-- extend bigistore should correctly extend internal aggregation state while summing istores which total length is larger than default capacity;
-- ./spec/istore/aggregates_spec.rb:32;
CREATE EXTENSION istore;
CREATE TABLE test (v bigistore);
INSERT INTO test SELECT bigistore(array_agg(x), array_agg(x)) FROM generate_series(1, 31, 2) AS a(x);
INSERT INTO test SELECT bigistore(array_agg(x), array_agg(x)) FROM generate_series(2, 32, 2) AS a(x);
SELECT SUM(v) FROM test;
                                                                                                                                                                                     sum                                                                                                                                                                                      
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 "1"=>"1", "2"=>"2", "3"=>"3", "4"=>"4", "5"=>"5", "6"=>"6", "7"=>"7", "8"=>"8", "9"=>"9", "10"=>"10", "11"=>"11", "12"=>"12", "13"=>"13", "14"=>"14", "15"=>"15", "16"=>"16", "17"=>"17", "18"=>"18", "19"=>"19", "20"=>"20", "21"=>"21", "22"=>"22", "23"=>"23", "24"=>"24", "25"=>"25", "26"=>"26", "27"=>"27", "28"=>"28", "29"=>"29", "30"=>"30", "31"=>"31", "32"=>"32"
(1 row)

ROLLBACK;
BEGIN;
-- extend istore should correctly aggregate using sum_floor;
-- ./spec/istore/aggregates_spec.rb:49;
CREATE EXTENSION istore;
CREATE TABLE test (v istore);
INSERT INTO test SELECT istore(array_agg(x), array_agg(x)) FROM generate_series(1, 21, 2) AS a(x);
INSERT INTO test SELECT istore(array_agg(x), array_agg(x)) FROM generate_series(2, 22, 2) AS a(x);
INSERT INTO test SELECT istore(array_agg(x), array_agg(x)) FROM generate_series(1, 21, 2) AS a(x);
SELECT SUM_FLOOR(v, 10) FROM test;
                                                                         sum_floor                                                                          
------------------------------------------------------------------------------------------------------------------------------------------------------------
 "10"=>"10", "11"=>"22", "12"=>"12", "13"=>"26", "14"=>"14", "15"=>"30", "16"=>"16", "17"=>"34", "18"=>"18", "19"=>"38", "20"=>"20", "21"=>"42", "22"=>"22"
(1 row)

ROLLBACK;
BEGIN;
-- extend bigistore should correctly aggregate using sum_floor;
-- ./spec/istore/aggregates_spec.rb:49;
CREATE EXTENSION istore;
CREATE TABLE test (v bigistore);
INSERT INTO test SELECT bigistore(array_agg(x), array_agg(x)) FROM generate_series(1, 21, 2) AS a(x);
INSERT INTO test SELECT bigistore(array_agg(x), array_agg(x)) FROM generate_series(2, 22, 2) AS a(x);
INSERT INTO test SELECT bigistore(array_agg(x), array_agg(x)) FROM generate_series(1, 21, 2) AS a(x);
SELECT SUM_FLOOR(v, 10) FROM test;
                                                                         sum_floor                                                                          
------------------------------------------------------------------------------------------------------------------------------------------------------------
 "10"=>"10", "11"=>"22", "12"=>"12", "13"=>"26", "14"=>"14", "15"=>"30", "16"=>"16", "17"=>"34", "18"=>"18", "19"=>"38", "20"=>"20", "21"=>"42", "22"=>"22"
(1 row)

ROLLBACK;
